src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,193 @@
+    private MongoCollection<Document> collection;
+    private CacheToStoreConverter<K, V> cacheToStoreConverter;
+
+
+    @Override
+    public Set<Characteristic> characteristics() {
+        return EnumSet.of(Characteristic.SHAREABLE, Characteristic.BULK_READ, Characteristic.EXPIRATION);
+    }
+
+
+    @Override
+    public CompletionStage<Void> start(InitializationContext ctx) {
+        this.context = ctx;
+
+        MongoDbStoreConfiguration configuration = ctx.getConfiguration();
+        ConnectionConfiguration connectionConfiguration = configuration.connection();
+
+        this.cacheToStoreConverter = configuration.converter() == JSON
+                ? new JsonCacheToStoreConverter<>(ctx, ctx.getPersistenceMarshaller())
+                : new BinaryCacheToStoreConverter<>(ctx, ctx.getPersistenceMarshaller());
+
+        String connectionUri = connectionConfiguration.uri();
+        ConnectionString connString = new ConnectionString(connectionUri);
+        MongoClientSettings settings = MongoClientSettings.builder()
+                .applyConnectionString(connString)
+                .retryWrites(true)
+                .build();
+        this.mongoClient = MongoClients.create(settings);
+
+        String databaseName;
+        if (connString.getDatabase() != null) {
+            databaseName = connString.getDatabase();
+        } else {
+            databaseName = connectionConfiguration.database();
+        }
+        String collectionName = connectionConfiguration.collection();
+        MongoDatabase database = mongoClient.getDatabase(databaseName);
+        this.collection = database.getCollection(collectionName);
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public CompletionStage<Void> stop() {
+        mongoClient.close();
+        return CompletableFuture.completedFuture(null);
+    }
+
+
+    @Override
+    public CompletionStage<Long> size(IntSet segments) {
+        return Single.fromPublisher(collection.countDocuments()).toCompletionStage();
+    }
+
+    @Override
+    public CompletionStage<Long> approximateSize(IntSet segments) {
+        return size(segments);
+    }
+
+
+    @Override
+    public CompletionStage<Boolean> containsKey(int segment, Object key) {
+        return load(segment, key).thenApply(Objects::nonNull);
+    }
+
+    @Override
+    public CompletionStage<MarshallableEntry<K, V>> load(int segment, Object key) {
+        long now = context.getTimeService().wallClockTime();
+        Object storeKey = cacheToStoreConverter.toStoreKey(key);
+        Bson filter = and(eq("_id", storeKey), or(gt(EXPIRY_TIME, now), lte(EXPIRY_TIME, -1L)));
+
+        return Maybe.fromPublisher(collection.find(filter))
+                .map(document -> cacheToStoreConverter.toCacheEntry(key, document))
+                .toCompletionStage(null);
+    }
+
+    @Override
+    public CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> marshallableEntry) {
```

Comment:
It looks like the `batch` command was not overridden. It looks like the mongo client is perfect for this method as well. Can we add that?

The `batch` method is used with ISPN transactions, batch API and when configured as an async store. Note that if the mongo client recommends having a bunch of smaller batch to us the `StoreConfiguration#maxBatchSize` configuration to determine how large of a batch to send to the mongo server.

Reply:
You can use the `bulkWrite` method where a batch is sent via using `Flowable.buffer(context.getConfiguration().maxBatchSize())` to create the batches. Just make sure to use concatMap when creating the sub Publishers so we only have one outstanding batch a time.

Reply:
OK, I'm going to add the `batch` method.

Reply:
Sorry, I'm not familiar with RxJava, and I cannot make it work with `concatMap` (instead of `concatMapEager`), unless I explicitly handle backpressure.

With this:
```java
        Flowable<WriteModel<Document>> entriesWritten = Flowable.fromPublisher(writePublisher)
                .concatMap(Flowable::fromPublisher)
                .map(e -> {
                    var storeKey = cacheToStoreConverter.toStoreKey(e.getKey());
                    return new ReplaceOneModel<>(eq("_id", storeKey), cacheToStoreConverter.toStoreEntry(storeKey, e), new ReplaceOptions().upsert(true));
                });

        Flowable<WriteModel<Document>> removedKeys = Flowable.fromPublisher(removePublisher)
                .concatMap(Flowable::fromPublisher)
                .map(key -> {
                    var storeKey = cacheToStoreConverter.toStoreKey(key);
                    return new DeleteOneModel<>(eq("_id", storeKey));
                });

        return Flowable.concatArrayEager(entriesWritten, removedKeys)
                .buffer(context.getConfiguration().maxBatchSize())
                .concatMap(writeModels -> Flowable.fromPublisher(collection.bulkWrite(writeModels)))
                .ignoreElements()
                .toCompletionStage(null);
```
`testWriteAndDeleteBatch` fails with:
> io.reactivex.rxjava3.exceptions.MissingBackpressureException: Unable to emit a new group (`#2`) due to lack of requests. Please make sure the downstream can always accept a new group as well as each group is consumed in order for the whole operator to be able to proceed.

Any ideas?

Reply:
Yeah, the concatMap here only subscribes to a single publisher at a time. You need to use `concatMapEager(Function, publisherCount, publisherCount)` on both the write and remove `Publisher<SegmentedPublisher>`s since there are multiples generated from the `groupBy` operator.

Note the default impl does something very similar on the interface as well.

Reply:
Yes, sorry, I thought you meant to use exactly `concatMap` (not `*Eager`) in the two branches, since I've seen from the `batch` doc that only the `Flowable.concatArrayEager` is required for the two input publishers.

But indeed (now I had the time to look a bit deeper) also the groups require eager subscription.

Thanks